File: backend/schemas/data_plane.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any, Optional

from pydantic import BaseModel, Field, conint

from .common import (
    ImportantEventType,
    KpiStatus,
    LifecycleEventType,
    MarketCode,
    RecoveryDrillCadence,
    RecoveryDrillStatus,
    RecoveryDrillTriggerMode,
    ReplayStatus,
    RequestModel,
    ScheduledIngestionStatus,
    TimestampSourceClass,
)


class DataIngestionRequest(RequestModel):
    symbol: str
    market: MarketCode = "TW"
    years: conint(ge=1) = 5  # type: ignore[valid-type]
    date_str: Optional[str] = None


class IngestionStageSummary(BaseModel):
    raw_payload_id: Optional[int] = None
    archive_object_reference: Optional[str] = None
    parser_version: Optional[str] = None
    input_rows: int
    validated_rows: int
    dropped_rows: int
    duplicates_removed: int
    null_rows_removed: int
    invalid_rows_removed: int
    gap_warnings: int
    upserted_rows: int
    official_overrides: int


class DataIngestionResponse(BaseModel):
    symbol: str
    market: str
    backfill: IngestionStageSummary
    daily_update: IngestionStageSummary


class ReplayRequest(RequestModel):
    raw_payload_id: int = Field(..., ge=1)
    benchmark_profile_id: Optional[str] = None
    notes: Optional[str] = None


class ReplayResponse(BaseModel):
    id: int
    raw_payload_id: int
    source_name: str
    symbol: str
    market: str
    archive_object_reference: Optional[str] = None
    parser_version: str
    benchmark_profile_id: Optional[str] = None
    notes: Optional[str] = None
    restore_status: ReplayStatus
    abort_reason: Optional[str] = None
    restored_row_count: int
    replay_started_at: datetime
    replay_completed_at: Optional[datetime] = None
    created_at: datetime


class RecoveryDrillRequest(RequestModel):
    raw_payload_id: Optional[int] = Field(default=None, ge=1)
    benchmark_profile_id: Optional[str] = None
    notes: Optional[str] = None


class RecoveryDrillResponse(BaseModel):
    id: int
    raw_payload_id: Optional[int] = None
    replay_run_id: Optional[int] = None
    benchmark_profile_id: Optional[str] = None
    notes: Optional[str] = None
    status: RecoveryDrillStatus
    trigger_mode: RecoveryDrillTriggerMode = "manual"
    schedule_id: Optional[int] = None
    scheduled_for_date: Optional[date] = None
    latest_replayable_day: Optional[date] = None
    completed_trading_day_delta: Optional[int] = None
    abort_reason: Optional[str] = None
    drill_started_at: datetime
    drill_completed_at: Optional[datetime] = None
    created_at: datetime


class RecoveryDrillScheduleRequest(RequestModel):
    market: MarketCode = "TW"
    symbol: Optional[str] = None
    cadence: RecoveryDrillCadence = "monthly"
    day_of_month: conint(ge=1, le=31) = 1  # type: ignore[valid-type]
    timezone: str = "Asia/Taipei"
    benchmark_profile_id: str
    notes: Optional[str] = None


class RecoveryDrillScheduleResponse(BaseModel):
    id: int
    market: MarketCode = "TW"
    symbol: Optional[str] = None
    cadence: RecoveryDrillCadence
    day_of_month: int
    timezone: str
    benchmark_profile_id: str
    notes: Optional[str] = None
    is_active: bool
    created_at: datetime


class BenchmarkProfileRequest(RequestModel):
    id: str
    cpu_class: str
    memory_size: str
    storage_type: str
    compression_settings: str
    archive_layout_version: str
    network_class: str


class BenchmarkProfileResponse(BaseModel):
    id: str
    cpu_class: str
    memory_size: str
    storage_type: str
    compression_settings: str
    archive_layout_version: str
    network_class: str
    created_at: datetime


class IngestionWatchlistRequest(RequestModel):
    symbol: str
    market: MarketCode = "TW"
    years: conint(ge=1) = 5  # type: ignore[valid-type]


class IngestionWatchlistResponse(BaseModel):
    id: int
    symbol: str
    market: MarketCode = "TW"
    years: int
    is_active: bool
    created_at: datetime


class ScheduledIngestionAttemptResponse(BaseModel):
    id: int
    attempt_number: int
    status: ScheduledIngestionStatus
    raw_payload_id: Optional[int] = None
    error_message: Optional[str] = None
    started_at: datetime
    completed_at: Optional[datetime] = None
    created_at: datetime


class ScheduledIngestionRunResponse(BaseModel):
    id: int
    watchlist_id: int
    symbol: str
    market: MarketCode = "TW"
    scheduled_for_date: date
    status: ScheduledIngestionStatus
    attempt_count: int
    last_error_message: Optional[str] = None
    first_attempt_at: Optional[datetime] = None
    last_attempt_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    attempts: list[ScheduledIngestionAttemptResponse] = Field(default_factory=list)


class IngestionDispatchRequest(RequestModel):
    scheduled_for_date: Optional[date] = None


class IngestionDispatchResponse(BaseModel):
    schedule_count: int
    dispatched_count: int
    succeeded_count: int
    failed_count: int
    skipped_count: int
    records: list[ScheduledIngestionRunResponse] = Field(default_factory=list)


class OpsKpiMetricResponse(BaseModel):
    value: Optional[float] = None
    status: KpiStatus
    numerator: Optional[float] = None
    denominator: Optional[float] = None
    unit: Optional[str] = None
    window: str
    details: dict[str, Any] = Field(default_factory=dict)


class OpsKpiResponse(BaseModel):
    gate_id: str
    overall_status: KpiStatus
    metrics: dict[str, OpsKpiMetricResponse]


class CrawlerRunResponse(BaseModel):
    source_name: str
    raw_payload_id: Optional[int] = None
    processed_count: int
    upserted_count: int
    errors: list[str] = Field(default_factory=list)


class LifecycleRecordUpsert(RequestModel):
    symbol: str
    market: MarketCode = "TW"
    event_type: LifecycleEventType
    effective_date: date
    reference_symbol: Optional[str] = None
    source_name: str
    raw_payload_id: Optional[int] = Field(default=None, ge=1)
    archive_object_reference: Optional[str] = None
    notes: Optional[str] = None


class ImportantEventUpsert(RequestModel):
    symbol: str
    market: MarketCode = "TW"
    event_type: ImportantEventType
    effective_date: Optional[date] = None
    event_publication_ts: datetime
    timestamp_source_class: TimestampSourceClass
    source_name: str
    raw_payload_id: Optional[int] = Field(default=None, ge=1)
    archive_object_reference: Optional[str] = None
    notes: Optional[str] = None


class LifecycleRecordResponse(BaseModel):
    id: int
    symbol: str
    market: MarketCode = "TW"
    event_type: LifecycleEventType
    effective_date: date
    reference_symbol: Optional[str] = None
    source_name: str
    raw_payload_id: Optional[int] = None
    archive_object_reference: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime


class ImportantEventResponse(BaseModel):
    id: int
    symbol: str
    market: MarketCode = "TW"
    event_type: ImportantEventType
    effective_date: Optional[date] = None
    event_publication_ts: datetime
    timestamp_source_class: TimestampSourceClass
    source_name: str
    raw_payload_id: Optional[int] = None
    archive_object_reference: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime

File: frontend/src/lib/types/dataPlane.ts
import type {
  ImportantEventType,
  KpiStatus,
  LifecycleEventType,
  MarketCode,
  RecoveryDrillCadence,
  RecoveryDrillTriggerMode,
  TimestampSourceClass,
} from "./common";

export interface IngestionStageSummary {
  raw_payload_id: number | null;
  archive_object_reference: string | null;
  parser_version: string | null;
  input_rows: number;
  validated_rows: number;
  dropped_rows: number;
  duplicates_removed: number;
  null_rows_removed: number;
  invalid_rows_removed: number;
  gap_warnings: number;
  upserted_rows: number;
  official_overrides: number;
}

export interface DataIngestionRequest {
  symbol: string;
  market: MarketCode;
  years: number;
  date_str?: string;
}

export interface DataIngestionResponse {
  symbol: string;
  market: string;
  backfill: IngestionStageSummary;
  daily_update: IngestionStageSummary;
}

export interface TwDailyReadinessRequest {
  market: "TW";
  symbols: string[];
  date_range?: {
    start: string;
    end: string;
  };
}

export interface TwDailyReadinessSymbolStatus {
  symbol: string;
  status: "ready" | "warning" | "missing";
  latest_daily_date: string | null;
  latest_raw_fetch_ts: string | null;
  requested_trading_days: number | null;
  covered_trading_days: number | null;
  missing_trading_days: number | null;
  stale_trading_days: number;
  warnings: string[];
}

export interface TwDailyReadinessResponse {
  market: "TW";
  overall_status: "ready" | "warning" | "missing";
  evaluated_at: string;
  date_range: {
    start: string;
    end: string;
  } | null;
  summary: {
    ready: number;
    warning: number;
    missing: number;
    stale: number;
  };
  symbols: TwDailyReadinessSymbolStatus[];
}

export interface ReplayRequest {
  raw_payload_id: number;
  benchmark_profile_id?: string;
  notes?: string;
}

export interface ReplayRecord {
  id: number;
  raw_payload_id: number;
  source_name: string;
  symbol: string;
  market: string;
  archive_object_reference: string | null;
  parser_version: string;
  benchmark_profile_id: string | null;
  notes: string | null;
  restore_status: "succeeded" | "failed";
  abort_reason: string | null;
  restored_row_count: number;
  replay_started_at: string;
  replay_completed_at: string | null;
  created_at: string;
}

export interface RecoveryDrillRequest {
  raw_payload_id?: number;
  benchmark_profile_id?: string;
  notes?: string;
}

export interface RecoveryDrillRecord {
  id: number;
  raw_payload_id: number | null;
  replay_run_id: number | null;
  benchmark_profile_id: string | null;
  notes: string | null;
  status: "succeeded" | "failed";
  trigger_mode: RecoveryDrillTriggerMode;
  schedule_id: number | null;
  scheduled_for_date: string | null;
  latest_replayable_day: string | null;
  completed_trading_day_delta: number | null;
  abort_reason: string | null;
  drill_started_at: string;
  drill_completed_at: string | null;
  created_at: string;
}

export interface RecoveryDrillScheduleRequest {
  market: MarketCode;
  symbol?: string;
  cadence?: RecoveryDrillCadence;
  day_of_month: number;
  timezone?: string;
  benchmark_profile_id: string;
  notes?: string;
}

export interface RecoveryDrillScheduleRecord {
  id: number;
  market: MarketCode;
  symbol: string | null;
  cadence: RecoveryDrillCadence;
  day_of_month: number;
  timezone: string;
  benchmark_profile_id: string;
  notes: string | null;
  is_active: boolean;
  created_at: string;
}

export interface LifecycleRecordPayload {
  symbol: string;
  market: MarketCode;
  event_type: LifecycleEventType;
  effective_date: string;
  reference_symbol?: string;
  source_name: string;
  raw_payload_id?: number;
  archive_object_reference?: string;
  notes?: string;
}

export interface LifecycleRecord extends LifecycleRecordPayload {
  id: number;
  created_at: string;
}

export interface ImportantEventPayload {
  symbol: string;
  market: MarketCode;
  event_type: ImportantEventType;
  effective_date?: string;
  event_publication_ts: string;
  timestamp_source_class: TimestampSourceClass;
  source_name: string;
  raw_payload_id?: number;
  archive_object_reference?: string;
  notes?: string;
}

export interface ImportantEvent extends ImportantEventPayload {
  id: number;
  created_at: string;
}

export interface OpsKpiMetricResponse {
  value: number | null;
  status: KpiStatus;
  numerator?: number | null;
  denominator?: number | null;
  unit?: string | null;
  window: string;
  details: Record<string, unknown>;
}

export interface OpsKpiResponse {
  gate_id: string;
  overall_status: KpiStatus;
  metrics: Record<string, OpsKpiMetricResponse>;
}

export interface ResearchGateArtifactResponse {
  status: KpiStatus;
  details: Record<string, unknown>;
}

export interface ResearchPhaseGateResponse {
  gate_id: string;
  verification_gate_id: string;
  overall_status: KpiStatus;
  metrics: Record<string, OpsKpiMetricResponse>;
  artifacts: Record<string, ResearchGateArtifactResponse>;
  missing_reasons: string[];
}
